helpers/focus.js

```diff
@@ -26,21 +26,34 @@ export function getComposedActiveElement() {
 	return node;
 }
 
-export function getFirstFocusableDescendant(node, includeHidden, predicate, includeTabbablesOnly) {
-	if (predicate === undefined) predicate = () => true;
-
+export function getFirstFocusableDescendant(node, includeHidden, predicate = () => true, includeTabbablesOnly, checkNodeVisibility = true) {
 	const composedChildren = getComposedChildren(node);
+	if (!composedChildren?.length || (!includeHidden && checkNodeVisibility && !isVisible(node))) return null;
 
 	for (let i = 0; i < composedChildren.length; i++) {
-		if (isFocusable(composedChildren[i], includeHidden, includeTabbablesOnly) && predicate(composedChildren[i])) return composedChildren[i];
+		if (!includeHidden && !isVisible(composedChildren[i], { checkAncestors: false })) continue;
+		if (isFocusable(composedChildren[i], true, includeTabbablesOnly) && predicate(composedChildren[i])) return composedChildren[i];
 
-		const focusable = getFirstFocusableDescendant(composedChildren[i], includeHidden, predicate, includeTabbablesOnly);
+		const focusable = getFirstFocusableDescendant(composedChildren[i], includeHidden, predicate, includeTabbablesOnly, false);
 		if (focusable) return focusable;
 	}
 
 	return null;
 }
 
+export function getFocusAlternative(node, includeHidden, predicate = () => true, includeTabbablesOnly, checkNodeVisibility = true) {
+	if (!node) return null;
+
+	if (!includeHidden && checkNodeVisibility) node = getFirstVisibleAncestor(node);
+
+	if (isFocusable(node, true) && predicate(node)) return node;
+
+	const focusableDescendant = getFirstFocusableDescendant(node, includeHidden, predicate, includeTabbablesOnly, false);
+	if (focusableDescendant !== null) return focusableDescendant;
+
+	return getFocusAlternative(getComposedParent(node), false);
```

Comment:
What's the difference between recusively calling `getFocusAlternative` here vs calling `getNextFocusable`?